dashboards-observability/public/components/explorer/explorer.tsx

```diff
@@ -108,49 +92,69 @@ export const Explorer = ({
   tabId,
   savedObjects,
   timestampUtils,
-  setToast
+  setToast,
 }: IExplorerProps) => {
-
   const dispatch = useDispatch();
-  const requestParams = { tabId, };
-  const {
-    isEventsLoading,
-    getEvents,
-    getAvailableFields
-  } = useFetchEvents({
+  const requestParams = { tabId };
+  const { isEventsLoading, getEvents, getAvailableFields } = useFetchEvents({
     pplService,
-    requestParams
+    requestParams,
   });
-  const {
-    isVisLoading,
-    getVisualizations,
-    getCountVisualizations
-  } = useFetchVisualizations({
+  const { isVisLoading, getVisualizations, getCountVisualizations } = useFetchVisualizations({
     pplService,
-    requestParams
+    requestParams,
   });
 
   const query = useSelector(selectQueries)[tabId];
   const explorerData = useSelector(selectQueryResult)[tabId];
   const explorerFields = useSelector(selectFields)[tabId];
   const countDistribution = useSelector(selectCountDistribution)[tabId];
   const explorerVisualizations = useSelector(selectExplorerVisualization)[tabId];
-  
+
   const [selectedContentTabId, setSelectedContentTab] = useState(TAB_EVENT_ID);
   const [selectedCustomPanelOptions, setSelectedCustomPanelOptions] = useState([]);
   const [selectedPanelName, setSelectedPanelName] = useState('');
   const [curVisId, setCurVisId] = useState('bar');
   const [prevIndex, setPrevIndex] = useState('');
-  const [isPanelTextFieldInvalid, setIsPanelTextFieldInvalid ] = useState(false);
+  const [isPanelTextFieldInvalid, setIsPanelTextFieldInvalid] = useState(false);
   const [isSidebarClosed, setIsSidebarClosed] = useState(false);
-  
+  const [timeIntervalOptions, setTimeIntervalOptions] = useState(TIME_INTERVAL_OPTIONS);
+
   const queryRef = useRef();
   const selectedPanelNameRef = useRef();
   const explorerFieldsRef = useRef();
   queryRef.current = query;
   selectedPanelNameRef.current = selectedPanelName;
   explorerFieldsRef.current = explorerFields;
 
+  let minInterval = 'y';
+  const findAutoInterval = (startTime: string, endTime: string) => {
+    if (startTime?.length === 0 || endTime?.length === 0 || startTime === endTime) return 'd';
+    const momentStart = dateMath.parse(startTime)!;
+    const momentEnd = dateMath.parse(endTime)!;
+    const diffSeconds = momentEnd.unix() - momentStart.unix();
+
+    // less than 1 second
+    if (diffSeconds <= 1) minInterval = 'ms';
+    // less than 2 minutes
+    else if (diffSeconds <= 60 * 2) minInterval = 's';
+    // less than 2 hours
+    else if (diffSeconds <= 3600 * 2) minInterval = 'm';
+    // less than 2 days
+    else if (diffSeconds <= 86400 * 2) minInterval = 'h';
+    // less than 1 month
+    else if (diffSeconds <= 86400 * 31) minInterval = 'd';
+    // less than 3 months
+    else if (diffSeconds <= 86400 * 93) minInterval = 'w';
+    // less than 1 year
+    else if (diffSeconds <= 86400 * 366) minInterval = 'M';
```

Comment:
question, this looks to be from trace analytics. I see discover can also have interval per multiple of a time unit (e.g. 3 hours, 12 hours), is that possible in observability?

Reply:
This is inspired from trace-analytics 💯 . This code deals with the the interval options for the first request made to the PPL backend. Hence, it only uses time bucket from datepicker and assigns the best one to Auto.  

The dynamic interval part from discover, is populated post response (i.e. after the first response). For example, if the response buckets are in hours then the new interval option of 30 minutes is added. This part is not added to observability yet, we should add it later.